include/swift/AST/LocalizationFormat.h

```diff
@@ -28,6 +30,92 @@ namespace swift {
 enum class DiagID : uint32_t;
 
 namespace diag {
+using namespace llvm::support;
+
+class LocalizationTableInfo {
+public:
+  using key_type = llvm::StringRef;
+  using key_type_ref = key_type;
+  using data_type = llvm::StringRef;
+  using data_type_ref = data_type;
+  using hash_value_type = uint32_t;
+  using offset_type = uint32_t;
+
+  hash_value_type ComputeHash(key_type_ref key) { return llvm::djbHash(key); }
+
+  std::pair<unsigned, unsigned> EmitKeyDataLength(llvm::raw_ostream &out,
+                                                  key_type_ref key,
+                                                  data_type_ref data) {
+    uint32_t keyLength = (uint32_t)key.size();
+    uint32_t dataLength = (uint32_t)data.size();
+    endian::Writer writer(out, little);
+    writer.write<uint32_t>(keyLength);
+    writer.write<uint32_t>(dataLength);
+    return {keyLength, dataLength};
+  }
+
+  void EmitKey(llvm::raw_ostream &out, key_type_ref key, unsigned len) {
+    out << key;
+  }
+
+  void EmitData(llvm::raw_ostream &out, key_type_ref key, data_type_ref data,
+                unsigned len) {
+    out << data;
+  }
+};
+
+class LocalizationLookupInfo {
+public:
+  using internal_key_type = llvm::StringRef;
+  using external_key_type = internal_key_type;
+  using data_type = llvm::StringRef;
+  using hash_value_type = uint32_t;
+  using offset_type = uint32_t;
+
+  internal_key_type GetInternalKey(external_key_type key) { return key; }
+
+  external_key_type GetExternalKey(internal_key_type key) { return key; }
+
+  static bool EqualKey(internal_key_type lhs, internal_key_type rhs) {
+    return lhs == rhs;
+  }
+
+  hash_value_type ComputeHash(internal_key_type key) {
+    return llvm::djbHash(key);
+  }
+
+  static std::pair<unsigned, unsigned> ReadKeyDataLength(const uint8_t *&data) {
+    unsigned keyLength = endian::readNext<uint32_t, little, unaligned>(data);
+    unsigned dataLength = endian::readNext<uint32_t, little, unaligned>(data);
+    return {keyLength, dataLength};
+  }
+
+  internal_key_type ReadKey(const uint8_t *data, offset_type length) {
+    return internal_key_type((const char *)data, length);
+  }
+
+  data_type ReadData(llvm::StringRef Key, const uint8_t *data,
+                     offset_type length) {
+    return data_type((const char *)data, length);
+  }
+};
+
+class SerializedLocalizationWriter {
+  llvm::OnDiskChainedHashTableGenerator<LocalizationTableInfo> generator;
+  std::error_code error;
+  llvm::raw_fd_ostream OS;
+
+public:
+  SerializedLocalizationWriter(llvm::StringRef filePath)
```

Comment:
Let's re-design interface here a little bit - make it:

1. `SerializedLocalizationWriter` should only have `generator` as a field;
2. Constructor could be removed - it's incorrect to mutate file before any data is added to the hash table.

3. Let's separate logic from destructor into a separate `emit` method.
```
void insert(llvm::StringRef id, llvm::StringRef translation);
bool emit(llvm::StringRef filePath) const;
```

4. Destructor could be removed.